src/auth/Crypto.cc

```diff
@@ -353,19 +365,405 @@ CryptoKeyHandler *CryptoAES::get_key_handler(const bufferptr& secret,
+  static constexpr size_t IV_SIZE = 128 / 8;
+  static constexpr size_t BLOCK_SIZE = 128 / 8;
+};
+
+
+class CryptoAES_256_CBCKeyHandler : public CryptoKeyHandler {
+public:
+  CryptoAES_256_CBCKeyHandler(const bufferptr& key, const bufferptr& iv)
+    : iv(iv) {
+    secret = key;
+  }
+  virtual ~CryptoAES_256_CBCKeyHandler() {}
+
+  int encrypt(const bufferlist& in, bufferlist& out, std::string *error) const {
+    return cbc_transform(in, out, true, error);
+  }
+  int decrypt(const bufferlist& in, bufferlist& out, std::string *error) const {
+    return cbc_transform(in, out, false, error);
+  }
+private:
+  bufferptr iv;
+  int cbc_transform(const bufferlist& in, bufferlist& out,
+      bool encrypt, std::string *error) const;
+  int cbc_transform(
+      const unsigned char* in, size_t size, unsigned char* out,
+      bool encrypt, std::string *error) const ;
+};
+
+int CryptoAES_256_CBCKeyHandler::cbc_transform(
+    const bufferlist& in, bufferlist& out,
+    bool encrypt, std::string *error) const
+{
+  if ((in.length() % CryptoAES_256_CBC::BLOCK_SIZE) != 0) {
+      return -EINVAL;
+    }
+  int result;
+  bufferptr p(in.length());
+  bufferlist _in(in);
+  result = cbc_transform(
+      reinterpret_cast<const unsigned char*>(_in.c_str()),
+      in.length(),
+      reinterpret_cast<unsigned char*>(p.c_str()),
+      encrypt, error);
+  out.append(p);
+  return result;
+}
+
+#ifdef USE_CRYPTOPP
 
+int CryptoAES_256_CBCKeyHandler::cbc_transform(
+    const unsigned char* in, size_t size, unsigned char* out,
+    bool encrypt, std::string *error) const
+{
+  using namespace CryptoPP;
+  if (encrypt) {
+    CBC_Mode< AES >::Encryption e;
+    e.SetKeyWithIV(
+        reinterpret_cast<const byte*>(secret.c_str()), CryptoAES_256_CBC::KEY_SIZE,
+        reinterpret_cast<const byte*>(iv.c_str()), CryptoAES_256_CBC::IV_SIZE);
+    e.ProcessData((byte*)out, (byte*)in, size);
+  } else {
+    CBC_Mode< AES >::Decryption d;
+    d.SetKeyWithIV(
+        reinterpret_cast<const byte*>(secret.c_str()), CryptoAES_256_CBC::KEY_SIZE,
+        reinterpret_cast<const byte*>(iv.c_str()), CryptoAES_256_CBC::IV_SIZE);
+    d.ProcessData((byte*)out, (byte*)in, size);
+  }
+  return true;
+}
 
+#elif defined(USE_NSS)
+
+int CryptoAES_256_CBCKeyHandler::cbc_transform(
+    const unsigned char* in, size_t size, unsigned char* out,
+    bool encrypt, std::string *error) const
+{
+  int result = -EINVAL;
+  PK11SlotInfo *slot;
+  SECItem keyItem;
+  PK11SymKey *symkey;
+  CK_AES_CBC_ENCRYPT_DATA_PARAMS ctr_params = {0};
+  SECItem ivItem;
+  SECItem *param;
+  SECStatus ret;
+  PK11Context *ectx;
+  int written;
+
+  slot = PK11_GetBestSlot(CKM_AES_CBC, NULL);
+  if (slot) {
+    keyItem.type = siBuffer;
+    keyItem.data = reinterpret_cast<unsigned char*>(const_cast<char*>(secret.c_str()));
+    keyItem.len = CryptoAES_256_CBC::KEY_SIZE;
+    symkey = PK11_ImportSymKey(slot, CKM_AES_CBC, PK11_OriginUnwrap, CKA_UNWRAP, &keyItem, NULL);
```

Comment:
`CryptoAESKeyHandler::init()` makes these calls to `PK11_GetBestSlot()` and `PK11_ImportSymKey()` just once, instead of every call to `encrypt/decrypt()`. can the new key handlers do that too?